### PR TITLE
Introduce secure share

### PR DESCRIPTION
**Pull request by matszczygiel**
This implements the secure share. The authentication is performed on server demand, which responds with HTTP 401 and random nonce in the header. Then the client is expected to send the authentication ticket in the next request.

The key management implementation required the change in `norddrop_new()` function signature, thus the avalanche of changes in the bindings and test framework. I think it's best to review this PR commit by commit to follow the "avalanche".

Please, take a close look at the swig include files and overall bindings, I'm not 100% sure if I did it right.

There is one unresolved issue, namely what happens when upon sending auth ticket, the client receives HTTP 401. Should the client stop and report failure to the app or continue sending request with version downgraded ?

### CODE REVIEW

File: test/drop_test/ffi.py [hunk 1]
```diff
@@ -11,6 +11,11 @@
 
 DEBUG_PRINT_EVENT = True
 
+# fmt: off
+PRIV_KEY = bytes([164, 70, 230, 247, 55, 28, 255, 147, 128, 74, 83, 50, 181, 222, 212, 18, 178, 162, 242, 102, 220, 203, 153, 161, 142, 206, 123, 188, 87, 77, 126, 183])
+PUB_KEY = bytes([68, 103, 21, 143, 132, 253, 95, 17, 203, 20, 154, 169, 66, 197, 210, 103, 56, 18, 143, 142, 142, 47, 53, 103, 186, 66, 91, 201, 181, 186, 12, 136])
+# fmt: on
+
```
Comment (LukasPukenis) [inline test/drop_test/ffi.py:18]: This currently uses the same private/public pair for all the peers so it doesn't fully exercise that the signing actually works. It should be several separate keypairs for each runner if I got it right. Keypairs could be hardcoded in `test/run.py` for each runner

**Review by LukasPukenis — APPROVED** [review]
Very nice job.

I suggest adding a handler for `401` error and then not doing backward compatibility as:
- if we have v3 and the peer has a lower version then it's going to be another error and not 401
- if we have an older version and the peer has v3 then it will respond with 401 but we're unaware and have no handling for it except try a lower version
- 
So it looks like we have a pretty good detection if both peers actually have v3 and auth fails

File: drop-transfer/src/ws/client/mod.rs [hunk 2]
```diff
@@ -127,6 +128,65 @@ async fn establish_ws_conn(
...
+                let val = resp
+                    .headers()
+                    .get(drop_auth::http::WWWAuthenticate::KEY)
+                    .context("Missing 'www-authenticate' header")?
+                    .to_str()?;
+
+                let resp = drop_auth::http::WWWAuthenticate::parse(val)
+                    .context("Failed to parse 'www-authenticate' header")?;
+
+                drop_auth::create_ticket(&keypair, resp).context("Failed to creaate auth ticket")
```
[the remark this answers is not in the document]
Comment (matszczygiel) [reply drop-transfer/src/ws/client/mod.rs]: Good catch :+1:

File: drop-transfer/src/ws/client/mod.rs [hunk 3]
```diff
@@ -111,7 +112,7 @@ async fn establish_ws_conn(
 
         let url = format!("ws://{}:{}/drop/{ver}", ip, drop_config::PORT);
 
-        match tokio_tungstenite::client_async(url, &mut socket).await {
+        match make_request(&mut socket, &url, logger).await {
             Ok(_) => break ver,
```
[the remark this answers is not in the document]
Comment (matszczygiel) [reply drop-transfer/src/ws/client/mod.rs:117]: It breaks from the `loop` https://github.com/NordSecurity/libdrop/pull/14/files#diff-96e0f9fcaf29ad2aeb641dbaa8fd12d960e9d440ed98ee212b2f830c2ea671f6L104 and assigns to `ver` the value of `var` (the names are duplicated in this case)

File: test/drop_test/ffi.py (hunk 1 above)
Comment (matszczygiel) [reply test/drop_test/ffi.py:18]: Yeah, that's a fair point. Will try to implement that

File: drop-transfer/src/auth.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+use std::net::IpAddr;
+
+use drop_auth::{Keypair, PublicKey};
+
+pub trait Context: Sync + Send + 'static {
+    fn own(&self) -> Keypair;
+    fn peer_public(&self, ip: IpAddr) -> Option<PublicKey>;
+}
```
[the remark this answers is not in the document]
Comment (matszczygiel) [reply drop-transfer/src/auth.rs]: Yeah, you're probably right that we should use the structure. Will try to change that but remember that peer public key should be retrieved by callback

File: drop-transfer/src/ws/server/mod.rs [hunk 5]
```diff
@@ -62,32 +67,38 @@ struct TmpFileState {
...
+    struct MissingAuth(IpAddr);
     impl warp::reject::Reject for MissingAuth {}
 
     #[derive(Debug)]
-    struct Unauthrorized;
+    struct Unauthrorized(IpAddr);
     impl warp::reject::Reject for Unauthrorized {}
 
     async fn handle_rejection(
+        nonces: &Mutex<BTreeMap<IpAddr, Nonce>>,
```
[the remark this answers is not in the document]
Comment (matszczygiel) [reply drop-transfer/src/ws/server/mod.rs:89]: There is no point in using `HashMap` if you have ordered keys. It's an overkill

File: norddrop/src/ffi/types.rs [hunk 6]
```diff
@@ -99,11 +105,34 @@ pub struct norddrop_logger_cb {
     pub cb: norddrop_logger_fn,
 }
 
+#[allow(non_camel_case_types)]
+/// Writes the peer's public key into the buffer of length 32.
+/// The peer is identifed by IPv4 address passed as octets (4 bytes), the
+/// special one 0.0.0.0 means the caller itself. Returns 0 on success and 1 on
+/// failure or missing key
```
[the remark this answers is not in the document]
Comment (LukasPukenis) [reply norddrop/src/ffi/types.rs:112]: that's fine, thanks for explanation

File: drop-transfer/src/ws/server/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (matszczygiel) [reply drop-transfer/src/ws/server/mod.rs:89]: Ok, I can change that

File: drop-transfer/src/ws/client/mod.rs [hunk 7]
```diff
@@ -127,6 +128,65 @@ async fn establish_ws_conn(
...
+                    .headers()
+                    .get(drop_auth::http::WWWAuthenticate::KEY)
+                    .context("Missing 'www-authenticate' header")?
+                    .to_str()?;
+
+                let resp = drop_auth::http::WWWAuthenticate::parse(val)
+                    .context("Failed to parse 'www-authenticate' header")?;
+
+                drop_auth::create_ticket(&keypair, resp).context("Failed to creaate auth ticket")
+            };
```
[the remark this answers is not in the document]
Comment (matszczygiel) [reply drop-transfer/src/ws/client/mod.rs]: Just to catch errors in one place and use the advantage of `?` in the closure

File: drop-transfer/src/auth.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (matszczygiel) [reply drop-transfer/src/auth.rs]: @LukasPukenis I've implemented it as a structure in the latest commit. Can you take a look and see if it's a better suit?

File: test/drop_test/ffi.py (hunk 1 above)
Comment (matszczygiel) [reply test/drop_test/ffi.py:19]: Done

**Pull request merged**